crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -665,9 +597,34 @@ where
         let mut transparent_tx_pool = shielded_tx.sapling_value_balance();
 
         // Check the validity of the keys and get the transfer data
-        let changed_balances =
+        let mut changed_balances =
             self.validate_state_and_get_transfer_data(keys_changed)?;
 
+        let masp_address_hash = TransparentAddress(<[u8; 20]>::from(
+            ripemd::Ripemd160::digest(sha2::Sha256::digest(
+                &Address::Internal(Masp).serialize_to_vec(),
+            )),
+        ));
+        if !verify_sapling_balancing_value(
```

Comment:
Is it ok to run this check only on the balance of the masp? Shouldn't we also check that shieldings/unshieldings happened from/to the addresses declared in the transparent bundle?

Reply:
I believe/hope that this should be okay, though a second pair of eyes would be appreciated...

So let `a` be the state changes being checked by the VP. `a` can be represented as a pair whose first element is all account balances before the `Tx` is applied, and whose second element is all account balances after the `Tx` is applied. Let `b` be the transformation induced by the inner `Transaction`.  `b` can be represented as a pair whose first element is all the transparent inputs of the `Transaction`, and whose second element is all the transparent outputs of the `Transaction`. Let `a=b`, `a-b`, `a+b` represent element-wise equality, subtraction, and addition respectively.

The old algorithm in `main` prevents malleability by essentially checking that `a=b`. I.e. `Transfer::source` has the same address as the `Transaction::vin`s, that `Transafer::amount` is the sum of the `Transaction::vin`s, that `Transfer::token` is the token of the `Transaction::vin`s, and so on.

This PR instead does the following. It splits the validation of the state changes into two parts: one part is the validation of state changes caused by the inner `Transaction`, the other part is the validation of state changes not caused by the inner `Transaction`. The idea is that is that if both parts pass the validation, then the entire `Tx` must pass the validation.

So algebraically we first check that `a>=b`, i.e. that the inputs and outputs of the inner `Transaction` do not exceed the account balances before or after the entire `Tx`'s application. Then we break the entire `Tx` into two components using the identity `a=b+(a-b)` and check `b` separately from `a-b`. `b` is essentially the MASP `Transaction`, so we check that its transparent transaction value pool is exactly 0, that each transparent input is signed/authorized, and all the other cryptographic checks. As for `a-b`, we essentially want to check that all the VPs would accept the state change described by `a-b` (which represents all state changes excluding those induced by the inner `Transaction`). Since I don't think the latter check is possible, this PR instead ensures that the accounts whose balances would decrease in `a-b` (i.e. those for which their balance in the pair's second element is less than their balance in the first element) have also signed the entire `Tx`.

So back to your question: instead of checking that unshieldings happen to the addresses given in the transparent bundles, we merely check that all accounts that whose balances go down in `a-b` have signed off on the transaction. So if `a` represents an unshielding of 50 BTC to Alice and `b` represents an unshielding of 50 BTC to Alice, then `a-b=0` and no additional authorization/signing is required. But if an attacker embeds the `Transaction` inside a different `Tx` so that `a` is now an unshielding of 50 BTC to Bertha, then: `a-b` is a transfer of 50 BTC from Alice to Bertha (since `b+(a-b)=(unshield 50 BTC to Alice)+(transfer 50 BTC from Alice to Bertha)=(unshield 50 BTC to Bertha)=a`). Since Alice's balance goes down in `a-b`, the VP would check that the entire `Tx` has been signed/authorized by Alice (something an attacker would not be able to do, otherwise they would just steal the funds from Alice after she receives them) and reject everything.

Reply:
Thanks for this summary, it helps a lot! I have another question, say that we have a tx embedding two transfers:

1. An unshielding to Alice
2. A transparent transfer from Bob to Alice

More specifically the modifications in storage are the following (assume just one token, NAM):

| Address  | Pre  | Post  | Delta  |
|---|---|---|---|
| MASP  | 10  | 9  | -1  |
| Alice  | 0  | 4  | +4  |
| Bob  | 5  | 2  | -3  |

So the unshielding moved 1 NAM to Alice and Bob transferred 3 NAMs to Alice. The check on `a>=b` should be fine in this case (I guess). Now, the only negative diff is Bob's one so we require his signature on the tx. This is fine but now my question: could Bob redirect the unshielding from Alice to himself? So that the new changes in storage are:

| Address  | Pre  | Post  | Delta  |
|---|---|---|---|
| MASP  | 10  | 9  | -1  |
| Alice  | 0  | 3  | +3  |
| Bob  | 5  | 3  | -2  |

I guess the `a>=b` is still fine and the only negative diff is still Bob's one, so we require only Bob's signature which Bob can provide. Is this correct?

Reply:
Yes, only Bob's signature would be required. 2 NAM would be transferred from Bob's account to Alice's, and 1 NAM would be unshielded from the MASP to Alice. So from the perspective of the algorithm, Bob merely sent 1 less NAM to Alice - he did not appropriate/seize Alice's 1 NAM. Does this answer your question?

Reply:
Sounds good, one note, currently the replay protection mechanism has some logic based on the presence of valid or invalid signatures that we are not handling or native vps cause those usually don't check signatures, so we'll need to update that too